hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkTaskContextSupplier.java

```diff
@@ -90,4 +120,16 @@ public Option<String> getProperty(EngineProperty prop) {
     throw new HoodieException("Unknown engine property :" + prop);
   }
 
+  // This reader context is used to read records before write, like compaction, clustering.
+  @Override
+  public Option<HoodieReaderContext> getReaderContext(HoodieTableMetaClient metaClient, boolean useReaderContext) {
+    if (useReaderContext && parquetReaderOpt.isPresent()) {
+      return Option.of(new SparkFileFormatInternalRowReaderContext(
```

Comment:
I didn't see why the reader context is bind to the task context supplier.

Reply:
My suggestion is we just add another `HoodieCompactor` and choose the right one based on the given config:

```java
HoodieCompactors.getInstance(...)
```
might need to abstract a base class for the shared logic though.

Another choice is just adding a new method in `HoodieCompactor` and invoke it specifically in the file group reader code path(based on the read context there).